matchbox_server/src/state.rs

```diff
@@ -0,0 +1,134 @@
+use axum::{extract::ws::Message, Error};
+use matchbox_protocol::PeerId;
+use matchbox_signaling::{
+    common_logic::{self, StateObj},
+    SignalingError, SignalingState,
+};
+use serde::Deserialize;
+use std::{
+    collections::{HashMap, HashSet},
+    net::SocketAddr,
+};
+
+#[derive(Debug, Deserialize, Default, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct RoomId(pub String);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct RequestedRoom {
+    pub id: RoomId,
+    pub next: Option<usize>,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Peer {
+    pub uuid: PeerId,
+    pub room: RequestedRoom,
+    pub sender: tokio::sync::mpsc::UnboundedSender<std::result::Result<Message, Error>>,
+}
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct ServerState {
+    clients_waiting: StateObj<HashMap<SocketAddr, RequestedRoom>>,
+    clients_in_queue: StateObj<HashMap<PeerId, RequestedRoom>>,
+    clients: StateObj<HashMap<PeerId, Peer>>,
+    rooms: StateObj<HashMap<RequestedRoom, HashSet<PeerId>>>,
+}
+impl SignalingState for ServerState {}
+
+impl ServerState {
+    /// Add a waiting client to matchmaking
+    pub fn add_waiting_client(&mut self, origin: SocketAddr, room: RequestedRoom) {
+        self.clients_waiting.lock().unwrap().insert(origin, room);
+    }
+
+    /// Assign a peer id to a waiting client
+    pub fn assign_id_to_waiting_client(&mut self, origin: SocketAddr, peer_id: PeerId) {
+        let room = {
+            let mut lock = self.clients_waiting.lock().unwrap();
+            lock.remove(&origin).expect("waiting client")
+        };
+        {
+            let mut lock = self.clients_in_queue.lock().unwrap();
+            lock.insert(peer_id, room);
+        }
+    }
+
+    /// Remove the waiting peer, returning the peer's requested room
+    pub fn remove_waiting_peer(&mut self, peer_id: PeerId) -> RequestedRoom {
+        let room = {
+            let mut lock = self.clients_in_queue.lock().unwrap();
+            lock.remove(&peer_id).expect("waiting peer")
+        };
+        room
+    }
```

Comment:
These functions are thin wrappers around `mutex` operations, would it perhaps make more sense to wrap `ServerState` in `Arc<Mutex<_>>` instead of it's components?

Reply:
I can try that, and I noticed that myself.

Reply:
Not sure how but I just removed all the Arc<Mutex<T>> and I didn't need to do anything else. Not sure why they were there originally.